[WebXR][OpenXR] Implement WebXR Hand Input module

svillar opened the pull request:
#### 488518e096587a812b5432fe204a79b705794cc6
<pre>
[WebXR][OpenXR] Implement WebXR Hand Input module
<a href="https://bugs.webkit.org/show_bug.cgi?id=298036">https://bugs.webkit.org/show_bug.cgi?id=298036</a>

Reviewed by Adrian Perez de Castro.

This change adds support for the WebXR Hand Input module
(<a href="https://www.w3.org/TR/webxr-hand-input-1/)">https://www.w3.org/TR/webxr-hand-input-1/)</a> for ports using
OpenXR.

It basically uses an object called tracker that is able to
collect pose information about the hand joints. The OpenXR
API actually provide information for 1 joint more than the
WebXR spec (the palm joint) so we must omit it when submitting
joints information to WebXR.

Systems supporting the XR_EXT_hand_joints_motion_range
extension can make use of it and request hand tracking data
considering the full motion range of a human hand.

A new system properties object was added so that different
parts of the code can easily check the runtime capabilities
(for example the tracker object should not be created if the
runtime does not support hand tracking even if the extension
is supported, as it happens typically in desktops).

* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.cpp:
(WebKit::OpenXRExtensions::loadMethods):
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
(WebKit::OpenXRInput::create):
(WebKit::OpenXRInput::initialize):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
(WebKit::OpenXRInputSource::create):
(WebKit::OpenXRInputSource::initialize):
(WebKit::OpenXRInputSource::collectHandTrackingData const):
(WebKit::OpenXRInputSource::collectInputSource const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h:
(WebKit::systemProperties):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::createInstance):
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/299314@main">https://commits.webkit.org/299314@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e71511ae81338d02683892f87b842f2c70365285

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38270 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28921 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124766 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70648 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38966 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46852 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90005 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121542 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31032 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106325 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70509 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30087 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68423 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110707 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100470 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24627 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127827 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117103 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45496 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34322 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/98646 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45860 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102546 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98430 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25022 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43862 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21863 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41994 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45366 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51044 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145799 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44829 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37488 "Found 1 new JSC binary failure: testapi, Found 18565 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/ControlFlow/forInPrimitive.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default, ChakraCore.yaml/ChakraCore/test/Function/apply.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48176 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46516 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -165,6 +174,49 @@ XrResult OpenXRInputSource::suggestBindings(SuggestedBindings& bindings) const
+
+#if defined(XR_EXT_hand_joints_motion_range)
+    XrHandJointsMotionRangeInfoEXT motionRangeInfo = createOpenXRStruct<XrHandJointsMotionRangeInfoEXT, XR_TYPE_HAND_JOINTS_MOTION_RANGE_INFO_EXT>();
+    if (OpenXRExtensions::singleton().isExtensionSupported(XR_EXT_HAND_JOINTS_MOTION_RANGE_EXTENSION_NAME ""_span)) {
```
This will check for the extension every time that `::collectHandTrackingData()` runs&mdash;quite often in this case. I think we could cache a set of boolean flags in `OpenXRExtensions` for those extensions that are checked regularly. Given that the code is correct here, and the suggestion being an optimization, it's fine to leave it to a follow-up patch.

svillar replied:
Right, actually we do that in wolvic precisely to avoid that search on every frame. Will do that as it's an easy change.

aperezdc commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -418,6 +418,7 @@ if (ENABLE_WEBXR)
     SET_AND_EXPOSE_TO_BUILD(USE_OPENXR ${OpenXR_FOUND})
     SET_AND_EXPOSE_TO_BUILD(XR_USE_PLATFORM_EGL TRUE)
     SET_AND_EXPOSE_TO_BUILD(XR_USE_GRAPHICS_API_OPENGL_ES TRUE)
+    SET_AND_EXPOSE_TO_BUILD(ENABLE_WEBXR_HANDS TRUE)
```
Is there some reason to add this build-time option? I would image that if the OpenXR headers define the needed symbols we want to always build the support, because anyway it is checked at runtime whether hand tracking is available. Unless I am missing something, there should not be any downside to remove the option and only check for `#if defined(XR_EXT_hand_tracking)` in the code.

svillar replied:
We are not adding it :) we're just enabling it, the whole cross platform hand tracking code is guarded by that

aperezdc replied:
Oh, you're right, don't mind the comment then.

aperezdc reviewed (approved):
I left a small suggestion (which can be a follow-up patch), and a question, but otherwise the changes LGTM.

svillar committed:
[WebXR][OpenXR] Implement WebXR Hand Input module
https://bugs.webkit.org/show_bug.cgi?id=298036

Reviewed by Adrian Perez de Castro.

This change adds support for the WebXR Hand Input module
(https://www.w3.org/TR/webxr-hand-input-1/) for ports using
OpenXR.

It basically uses an object called tracker that is able to
collect pose information about the hand joints. The OpenXR
API actually provide information for 1 joint more than the
WebXR spec (the palm joint) so we must omit it when submitting
joints information to WebXR.

Systems supporting the XR_EXT_hand_joints_motion_range
extension can make use of it and request hand tracking data
considering the full motion range of a human hand.

A new system properties object was added so that different
parts of the code can easily check the runtime capabilities
(for example the tracker object should not be created if the
runtime does not support hand tracking even if the extension
is supported, as it happens typically in desktops).

* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.cpp:
(WebKit::OpenXRExtensions::loadMethods):
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
(WebKit::OpenXRInput::create):
(WebKit::OpenXRInput::initialize):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
(WebKit::OpenXRInputSource::create):
(WebKit::OpenXRInputSource::initialize):
(WebKit::OpenXRInputSource::collectHandTrackingData const):
(WebKit::OpenXRInputSource::collectInputSource const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h:
(WebKit::systemProperties):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::createInstance):
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/299314@main
